foundations/foundation-config/src/test/java/org/apache/servicecomb/config/inject/TestConfigObjectFactory.java

```diff
@@ -0,0 +1,437 @@
+    ArchaiusUtils.setProperty("booleanValueObj", true);
+    ArchaiusUtils.setProperty("booleanValueObj1", true);
+
+    Assert.assertEquals("strValue", config.strValue);
+    Assert.assertEquals("strValue1", config.getStrValue1());
+
+    Assert.assertEquals(1, config.intValue);
+    Assert.assertEquals(2, config.getIntValue1());
+    Assert.assertEquals(3, (int) config.intValueObj);
+    Assert.assertEquals(4, (int) config.getIntValueObj1());
+
+    Assert.assertEquals(5, config.longValue);
+    Assert.assertEquals(6, config.getLongValue1());
+    Assert.assertEquals(7, (long) config.longValueObj);
+    Assert.assertEquals(8, (long) config.getLongValueObj1());
+
+    Assert.assertEquals(9, config.floatValue, 0);
+    Assert.assertEquals(10, config.getFloatValue1(), 0);
+    Assert.assertEquals(11, config.floatValueObj, 0);
+    Assert.assertEquals(12, config.getFloatValueObj1(), 0);
+
+    Assert.assertEquals(13, config.doubleValue, 0);
+    Assert.assertEquals(14, config.getDoubleValue1(), 0);
+    Assert.assertEquals(15, config.doubleValueObj, 0);
+    Assert.assertEquals(16, config.getDoubleValueObj1(), 0);
+
+    Assert.assertTrue(config.booleanValue);
+    Assert.assertTrue(config.isBooleanValue1());
+    Assert.assertTrue(config.booleanValueObj);
+    Assert.assertTrue(config.getBooleanValueObj1());
+  }
+
+  @InjectProperties(prefix = "root")
+  public static class ConfigWithAnnotation {
+    @InjectProperty(prefix = "override", keys = {"high", "low"})
+    public String strValue;
+
+    @InjectProperty(keys = "${key}.value")
+    public int intValue;
+
+    @InjectProperty(keys = "${list-1}.a.${list-2}.b")
+    public long longValue;
+
+    @InjectProperty(keys = "${full-list}")
+    public float floatValue;
+
+    @InjectProperty(defaultValue = "abc")
+    public String strDef;
+
+    @InjectProperty(defaultValue = "1")
+    public int intDef;
+
+    @InjectProperty(defaultValue = "2")
+    public long longDef;
+
+    @InjectProperty(defaultValue = "3")
+    public float floatDef;
+
+    @InjectProperty(defaultValue = "4")
+    public double doubleDef;
+
+    @InjectProperty(defaultValue = "true")
+    public boolean booleanDef;
+  }
+
+  @Test
+  public void annotationDefault() {
+    ConfigWithAnnotation config = new ConfigObjectFactory().create(ConfigWithAnnotation.class);
+
+    Assert.assertEquals("abc", config.strDef);
+    Assert.assertEquals(1, config.intDef);
+    Assert.assertEquals(2, config.longDef);
+    Assert.assertEquals(3, config.floatDef, 0);
+    Assert.assertEquals(4, config.doubleDef, 0);
+    Assert.assertTrue(config.booleanDef);
+  }
+
+  @Test
+  public void placeholder_multi_list() {
+    ConfigWithAnnotation config = new ConfigObjectFactory().create(ConfigWithAnnotation.class,
+        "list-1", Arrays.asList("l1-1", "l1-2"),
+        "list-2", Arrays.asList("l2-1", "l2-2"));
+    // l1-1.a.l2-1.b
+    // l1-2.a.l2-1.b
+    // l1-1.a.l2-2.b
+    // l1-2.a.l2-2.b
```

Comment:
It seems that the backward placeholders have a higher priority. Does this meets your expectation?

Reply:
it's my expectation, but it's not correct
i'll fix it.

Reply:
done